Intro (#8)
Dlang-UPB/D-Summer-School: intro/README.md

# Introduction to D

## Syntax

The D programming language uses a C-style syntax that ensures a smooth transition for programmers coming from a C\C++ background. With no further ado, let us take a random C program and see what it takes to compile it with the D compiler:

> 

Reviewer: ```d
...cod
```

1. Imports may selectively specify which symbols are to be imported. For example, in the above code snippet, the full standard IO module of the standard library is imported, even though only the **printf** function is used.
This results in a degradation in compile time since there is a larger symbol pool that needs to be examined when trying to resolve a symbol.
In order to fix this, we can replace the blunt import with:
```d
import std.stdio : printf;
```

1. Imports may be used at any scope level.
```d
int main()
{
    int position = 7, c, n = 10;
    int[10] array = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

    for (c = position - 1; c < n - 1; c++)
        array[c] = array[c+1];

    // If the lines calling printf are deleted,
    // it is easier to spot the now useless import
    import std.stdio : printf;
    printf("Resultant array:\n");
    for (c = 0; c < n - 1; c++)
        printf("%d\n", array[c]);

    return 0;
}
```

### C to D

Go to this [link](https://github.com/TheAlgorithms/C/tree/master/searching). You will find a series of searching algorithms, each implemented in C in its own file.
- Choose one algorithm and port it to D with minimum modifications.

> - Update the code using D specific features to improve the code (fewer lines of code, increase in expressiveness etc.).

Reviewer: This should be done later. Right now we haven't shown anything about how powerful D is.

        int[] b = [1, 2, 3, 4, 5];
        a = 3;                     // set all elements of a to 3 
        b = 5;                     // set all elements of b to 5
        a[] = 2;                   // same as `a = 3`; using an empty slice is the same as slicing the full array
        b = a[0 .. $];             // $ evaluates to the length of the array (in this case 10)
        b = a[];                   // semantically equivalent to the one above
        b = a[0 .. a.length]       // semantically equivalent to the one above
        b[] = a[];                 // semantically equivalent to the one above
        b[2 .. 4] = 4              // same as b[2] = 4, b[3] = 4
        b[0 .. 4] = a[0 .. 4]      // error, a does not have 4 elements
        a[0 .. 3] = b              // error, operands have different length
    }
    ```
        
    **.length** is a builtin array property.
    For an extensive list of array properties click [here](https://dlang.org/spec/arrays.html#array-properties).

    Now let's see how arrays work with functions.
    The next quiz can be tricky but it will help us better understand how arrays work.

>     If it help, think what would happed in a similar situation in C.

Reviewer: ```suggestion
    If it helps, think what would happed in a similar situation in C.
```

>     [Quiz](./quiz/ref-vs-value.md)

Reviewer: This quiz requires knowledge about structs. Up until this point we haven't explained what structs are.

Reviewer: I suggest putting this quiz after we talk about structs and after we explain the fact that an array in D is seen as a struct that contains the pointer and a size.

    
2. Array Concatenation:

    ```d
    void main()
    {
        int[] a;
        int[] b = [1, 2, 3];
        int[] c = [4, 5];
     
        a = b ~ c;       // a will be [1, 2, 3, 4, 5]
        a = b;           // a refers to b
        a = b ~ c[0..0]; // a refers to a copy of b
        a ~= c;          // equivalent to a = a ~ c;
    }
    ```
       
    Concatenation always creates a copy of its operands, even if one of the operands is a 0-length array.

>     The operator `~=` does not always create a copy.

Reviewer: This is confusing. You should provide more information about this. When does `~=` create a copy and when does it not?

    For more information click [here](https://dlang.org/spec/arrays.html#resize).

    Navigate to **demo/slice** directory.
    Inspect and run the file **slice.d**.
    What happened here? Why?

>     [Quiz](./quiz/slice.md)

Reviewer: You could add here the exercise to improve the initial example by using slices instead of the for loop.

Author: If you are talking about the exercise with binary search where they have to fix a bug, I feel like the part with the unittest should be first, because chances are they will fix the bug without even realizing, and I think the main purpose of that exercise is to show how useful unittests can be

Reviewer: No, I was referring to the initial example where an element is deleted from an array.

    ```d
    void main()
    {
        int[string] aa;    
        aa["hello"] = 3;           
     
        aa.remove("hello")     // removes the pair ("hello", 3)
    }
    ```
    
    **remove(key)** does nothing if the given key does not exist and returns **false**.
    If the given key does exist, it removes it from the AA and returns **true**.
    All keys can be removed by using the method **clear**.
    
2. Testing membership:

    ```d
    void main()
    {
        int[string] aa;    
        aa["hello"] = 3;
     
        int* p;
        p = "hello" in aa;
     
        if (p)
        {
            *p = 4;  // update value associated with key; aa["hello"] == 4
        }
    }
    ```
    
    The “in” expression yields a pointer to the value if the key is in the associative array, or null if not.

    For more advanced operations on AAs check this [link](https://dlang.org/spec/hash-map.html#construction_assignment_entries).
    For an exhaustive list of the AA properties check this [link](https://dlang.org/spec/hash-map.html#properties).

Find the [majority element](https://leetcode.com/problems/majority-element/) in a string array using builtin associative arrays.

> You can start your implementation using the skeleton inside **majority-element** directory

Reviewer: ```suggestion
You can start your implementation using the skeleton inside **majority-element** directory.
```

## Structs

In D, structs are simple aggregations of data and their associated operations on that data:

```d
struct Rectangle
{
    size_t length, width;
    int id;
 
    size_t area() { return length*width; }
    size_t perimeter() { return 2*(length + width); }
    size_t isSquare() { return length == width; }
    void setId(int id) { this.id = id; }
}
```    

## Classes

In D, classes are very similar with java classes:

1. Classes can implement any number of interfaces

> 2. Classes can inherit a single class
> 3. Class objects are instantiated by reference only
> 4. super has the same meaning as in Java
> 5. overriding rules are very similar

Reviewer 2: ```suggestion
1. Classes can inherit a single class
1. Class objects are instantiated by reference only
1. super has the same meaning as in Java
1. overriding rules are very similar
```

The fundamental difference between **structs** and **classes** is that the former are **value** types, while the latter are **reference** types.
This means that whenever a struct is passed as an argument to an lvalue function parameter, the function will operate on a copy of the object.

> When a class is passed as an argument to an lvalue function parameter, the function will receive a reference to the object.

Reviewer 2: Does it make sense to explain what a reference is? I think 1 / 2 phrases that it's actually a pointer blah blah wouldn't hurt.

Both **structs** and **classes** will be covered more in depth in a further lab.

> Navigate to the **demo/struct-class** directory.

Reviewer 2: ```suggestion
Navigate to the `demo/struct-class` directory.
```

You will find 2 files implementing the same program.
One uses a struct as a data container, while the other uses a class.

- Compile both files and measure the run time of each program. How do you explain the differences?
- How do you explain the output differences? How we can change the code for structs so both versions produce the same output?

> [Quiz](./quiz/struct-class.md)

Reviewer: I think that the quiz ref-vs-value is much more suited here after they understand what a value type is. You can have any number of consecutive quizzes.

## Functions

Functions are declared the same as in C. In addition, D offers some convenience features like:

1. Uniform function call syntax (UFCS): allows that any call to a free function **fun(a)** can be written as member function call **a.fun()**

    ```d
    import std.algorithm : group;
    import std.range : chain, dropOne, front, retro;
    [1, 2].chain([3, 4]).retro; // 4, 3, 2, 1
    [1, 1, 2, 2, 2].group.dropOne.front; // (2, 3)
     
    front(dropOne(group([1, 1, 2, 2, 2])));
    ```

> D has a standard library called [phobos](https://dlang.org/phobos/).

Reviewer: This paragraph is very abrupt. You could maybe put the exercise first and then instruct them to search for functions  in the standard library. Something along the lines of: "Find the median [...]. Use any number of functions from the [standard library](link). Use UFCS to make the code more expressive."

Find the **median element** using functions from the [std.algorithm](https://dlang.org/phobos/std_algorithm.html) package.
Use UFCS for an increase in expressiveness.

> Observe the increase in performance achieved by using functions from the standard library

Reviewer: Is this a task? What are they supposed to do? Time the program?

Author: Not really, just run the program, in the skeleton they should complete 2 functions, 1 using the standard library, and one without. Then, inside the unittest there is an example and they are timed.

2. Overloading:

    ```d
    void fun(int a) {}
    void fun(int a, int b) {}
    ```
    

>     Overloading is done based on the **type of parameters**, not on the **return type**.

Reviewer: This section needs to be expanded. You should explain a bit what overloading is and also the example needs to contain some calls to these functions.

    
3. Default parameters:

    ```d
    void fun(int a, int b=8) {}
     
    void main()
    {
        fun(7);    // calls fun(7, 8)
        fun(2, 3); // calls fun(2, 3) 
    }
    ```
        
    A function may have any number of default parameters. If a default parameter is given, all following parameters must also have default values.
    
4. Auto return type:

    ```d
    auto fun()
    {
        return 7;
    }
     
    void main()
    {
        int b = fun();
        auto c = fun();   // works for variables also
    }
    ```
        
    Auto functions have their return type inferred based on the type of the return statements.
    Auto can be also used to infer the type of a variable declaration.
    

> Navigate to **demo/voldemort** directory. Inspect and run the code inside **voldermort.d**.

Reviewer 2: ```suggestion
Navigate to `demo/voldemort` directory. Inspect and run the code inside `voldermort.d`.
```

> Now navigate to **intro/voldemort** directory. 

Reviewer 2: ```suggestion
Now navigate to `intro/voldemort` directory. 
```

> **Result** is a struct declared inside **fun's** scope.

Reviewer 2: ```suggestion
`Result` is a struct declared inside `fun`'s scope.
```

> However we want to be able to use it in **main** as well.

Reviewer 2: ```suggestion
However, we want to be able to use it in `main` as well.
```

> Inspect the file. Does it compile? Why? Fix the issue.

Reviewer 2: ```suggestion
Inspect the file.
Does it compile?
Why?
Fix the issue.
```

    
## Unittests

Unit tests are a builtin framework of test cases applied to a module to determine if it is working properly.
A D program can be run with unit tests enabled or disabled.

Unit tests are a special function defined like:

```d
unittest
{
    // test code
}
```

Individual tests are specified in the unit test using [assert expressions](https://dlang.org/spec/expression.html#AssertExpression).
There can be any number of unit test functions in a module, including within struct and class declarations.
They are executed in lexical order.
Unit tests, when enabled, are run after all static initialization is complete and before the **main()** function is called.

```d
struct Sum
{
    int add(int x, int y) { return x + y; }
 
    unittest
    {
        Sum sum;
        assert(sum.add(3,4) == 7);
        assert(sum.add(-2,0) == -2);
    }
}
```

> Navigate to the **demo/binary-search** directory.

Reviewer 2: ```suggestion
Navigate to the `demo/binary-search` directory.
```

> Inspect the source file binarySearch.d.

Reviewer 2: ```suggestion
Inspect the source file `binarySearch.d`.
```

As the name implies, a binarySearch algorithm is implemented on integers.
Compile and run the file.

> - Write a unittest function that tests some corner cases. Are there any bugs in the algorithm implementation? If yes, fix them.

Reviewer 2: ```suggestion
- Write a unittest function that tests some corner cases.
Are there any bugs in the algorithm implementation?
If so, fix them.
```

- Rewrite the binarySearch algorithm to make use of slices.
   
## Contract Programming

> Contracts are a breakthrough technique to reduce the programming effort for large projects. Contracts are the concept of preconditions, postconditions and errors.

Reviewer 2: ```suggestion
Contracts are a breakthrough technique to reduce the programming effort for large projects.
Contracts are the concept of preconditions, postconditions and errors.
```

### Preconditions/Postconditions

> The pre contracts specify the preconditions before a statement is executed. The most typical use of this would be in validating the parameters to a function.

Reviewer 2: ```suggestion
The pre contracts specify the preconditions before a statement is executed.
The most typical use of this would be in validating the parameters to a function.
```

The expressions in the first are lowered to contract blocks that look almost exactly like the second, except that a separate block is created for each expression in the first, thus avoiding shadowing variable names.

[Quiz](./quiz/function-overloading.md)

Navigate to the **distribute-sum** directory.
Inspect the source file distribute.d and follow the instructions.

- Complete the `distribute` function.
- Write an `in` contract that ensures the value of the first argument(`sum`) is positive.
- Write an `out` contract that ensures the value of `sum` is equal to `first` + `second`.

### Invariants

Invariants are used to specify characteristics of a class or struct that must always be true (except while executing a member function).
For example, a class representing a date might have an invariant that the day must be 1..31 and the hour must be 0..23:

```d
class Date
{
    int day;
    int hour;
 
    this(int d, int h)
    {
        day = d;
        hour = h;
    }
 
    invariant
    {
        assert(1 <= day && day <= 31);
        assert(0 <= hour && hour < 24, "hour out of bounds");
    }
}
```

For public or exported functions, the order of execution is:
1. preconditions

> 2. invariant
> 3. function body
> 4. invariant
> 5. postconditions

Reviewer 2: ```suggestion
1. invariant
1. function body
1. invariant
1. postconditions
```

> Navigate to **my-associative-array** directory.

Reviewer 2: ```suggestion
Navigate to the `my-associative-array` directory.
```

> Follow the directions in the skeleton in order to implement our very own string : int associative array.

Reviewer 2: ```suggestion
Follow the directions in the skeleton in order to implement our very own `string : int` associative array.
```

     
        aa.remove("hello")     // removes the pair ("hello", 3)
    }
    ```
    
    `remove(key)` does nothing if the given key does not exist and returns `false`.
    If the given key does exist, it removes it from the AA and returns `true`.
    All keys can be removed by using the method `clear`.
    
2. Testing membership:

    ```d
    void main()
    {
        int[string] aa;    
        aa["hello"] = 3;
     
        int* p;
        p = "hello" in aa;
     
        if (p)
        {
            *p = 4;  // update value associated with key; aa["hello"] == 4
        }
    }
    ```
    
    The "i" expression yields a pointer to the value if the key is in the associative array, or `null` if not.

    For more advanced operations on AAs check this [link](https://dlang.org/spec/hash-map.html#construction_assignment_entries).
    For an exhaustive list of the AA properties check this [link](https://dlang.org/spec/hash-map.html#properties).

## Practice

Find the [majority element](https://leetcode.com/problems/majority-element/) in a string array using builtin associative arrays.
You can start your implementation using the skeleton inside `majority-element` directory.

> ### C to D

Reviewer: Shouldn't this also be tagged as "Practice" ?

    return 0;
}
```
    
As you can see, the only differences are:

- the include directive was replaced by an import statement;
- the array definition and initialization are slightly modified;

Most C programs require minimal changes in order to be compiled with the D compiler. So do not worry, even if you don't have any previous experience in D, you will be able to understand most of the programs written in it because the syntax is extremely similar to the C one.

The full D grammar can be found [here](https://dlang.org/spec/grammar.html).

Now, using the above code snippet, let us delve into the D specific concepts.

## Imports

In D, **imports** represent the counterpart of the C **include** directive, however the are 2 fundamental differences:

1. Imports may selectively specify which symbols are to be imported. For example, in the above code snippet, the full standard IO module of the standard library is imported, even though only the **printf** function is used. This results in a degradation in compile time since there is a larger symbol pool that needs to be examined when trying to resolve a symbol. In order to fix this, we can replace the blunt import with:

>         import std.stdio : printf;

Reviewer: use code block for this too.

2. Imports may be used at any scope level.
    ```d

>     int main()

Reviewer: this code is indented differently from the above snippets (it contains an extra tab)

The D programming language defines 3 classes of data types:

1. [Basic data types](https://dlang.org/spec/type.html#basic-data-types): such as int, float, long etc. that are similar to the ones provided by C;
2. [Derived data types](https://dlang.org/spec/type.html#derived-data-types): pointer, array, associative array, delegate, function;
3. [User defined types](https://dlang.org/spec/type.html#user-defined-types): class, struct, union, enum;

We will not insist on basic data types and pointers, as those are the same (or slightly modified versions) as the ones in C\C++. We will focus on arrays, associative arrays, classes, structs and unions. Delegates, functions and enums will be treated in a future lab.

Note that in D all types have a default value. This means that there are no uninitialized variables.

## Arrays

The fundamental difference between a C array and a D array is that the former is represented by a simple pointer, while the latter is represented by a pointer and a size. This design decision was taken because of the numerous cases of C buffer overflow attacks that can be simply mitigated by automatically checking the array bounds. Let's take a simple example:

> ```d

Reviewer: ```suggestion
```c
```

    #include <stdio.h>
    
    void main()
    {
        int a = 5;
        int b[10];
        b[11] = 9;
        printf("%d\n", a);
    }
```

> Compiling this code (without the stack protector activated) will lead to a buffer overflow in which the variable `a` will be modified and the print will output `9`. Compiling the same code with D (simply replace the include with an `import std.stdio : printf`) will issue a compile time error that states that 11 is beyond the size of the array. Aggregating the pointer of the array with its size facilitates a safer implementation of arrays.

Reviewer: One sentence per line please.

D implements 2 types of arrays:
1. [Static Arrays](https://dlang.org/spec/arrays.html#static-arrays): their length must be known at compile time and therefore cannot be modified; all examples up until this point have been using static arrays.
2. [Dynamic Arrays](https://dlang.org/spec/arrays.html#dynamic-arrays): their length may change dynamically at run time.

### Slicing

Slicing an array means to specify a subarray of it. An array slice does not copy the data, it is only another reference to it:

```d
void main()
{
    int[10] a;       // declare array of 10 ints
    int[] b;
 
    b = a[1..3];     // a[1..3] is a 2 element array consisting of a[1] and a[2]
    int x = b[1];    // equivalent to `int x = 0;`
    a[2] = 3;
    int y = b[1];    // equivalent to `int y = 3;`
}
```

### Quiz

What is the output of the following program?

```d
void main()
{
    import std.stdio : writeln;

    int[10] a;
    auto b = a[0 .. $];
    b = b ~ 5;

    writeln(b.length);

> }

Reviewer: This quiz is too early. We have not talked about the `$` and `~` operators.

```

- compilation error
- 11
- 10

### Array operations

1. Array setting:

    ```d
    void main()
    {
         int[3] a = [1, 2, 3];
         int[3] c = [ 1, 2, 3, 4];  // error: mismatched sizes
         int[] b = [1, 2, 3, 4, 5];
         a = 3;                     // set all elements of a to 3 
         b = 5;                     // set all elements of b to 5
         a[] = 2;                   // same as `a = 3`; using an empty slice is the same as slicing the full array
         b = a[0 .. $];             // $ evaluates to the length of the array (in this case 10)
         b = a[];                   // semantically equivalent to the one above
         b = a[0 .. a.length]       // semantically equivalent to the one above
         b[] = a[];                 // semantically equivalent to the one above
         b[2 .. 4] = 4              // same as b[2] = 4, b[3] = 4
         b[0 .. 4] = a[0 .. 4]      // error, a does not have 4 elements
         a[0 .. 3] = b              // error, operands have different length
    }
    ```
        

>     **.length** is a builtin array property. For an extensive list of array properties click [here](https://dlang.org/spec/arrays.html#array-properties).

Reviewer: Why is the indentation here different from the rest of the text?

    
    ```d
    int main()
    {
    int position = 7, c, n = 10; 
    int[] array = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
 
    array = array[0 .. position] ~ array[position + 1 .. $]; 
 
    import std.stdio;
    writeln("Resultant array:");
    writeln(array);                                                                                                                           
 
    return 0;
    }
    ```
        
    [writeln](https://dlang.org/library/std/stdio/writeln.html) is a function from the standard D library that does not require a format and is easily usable with a plethora of types.

    As you can see, the resulting code is much more expressive and fewer lines of code were utilized.

### Quiz
What is the output of the following code?

```d
void setLength(int[] a, size_t length)
{
    a.length = length;
}

void startWith2(int[] a)
{
    a[0] = 2;
    a[1] = 2;
}

void main()
{
    import std.stdio : writeln;
    int[] a = new int[5];
    
    a.setLength(2);
    writeln(a.length);

    a.startWith2();
    writeln(a);
}
```
    
- [2, 2]
- [2, 2, 0, 0, 0]
- [0, 0, 0, 0, 0]

> - [0, 0]

Reviewer: At this quiz you are also outputting the length, however the answers contain only the array.

### Exercise

> 1. C to D

Reviewer: I think that this exercise could be moved right after the import part, no?

Go to this [link](https://github.com/TheAlgorithms/C/tree/master/searching). You will find a series of searching algorithms, each implemented in C in its own file.
    - Choose one algorithm and port it to D with minimum modifications.
    - Update the code using D specific features to improve the code (fewer lines of code, increase in expressiveness etc.).

    ##### For 2 and 3 you can start your implementation using the skeleton inside **array-median** directory.

2. Compute the median element of an unsorted integer array. For this, you will have to:
    - Sort the array: implement any sorting algorithm you wish.
    - Eliminate the duplicates: once the array is sorted, eliminating the duplicates is trivial.
    - Check if the length of the array is **odd** or **even** and then return the median element.

> 3. D has a standard library called [phobos](https://dlang.org/phobos/). Implement exercise 2 using functions from the [std.algorithm](https://dlang.org/phobos/std_algorithm.html) package. Use UFCS for an increase in expressiveness. Observe the increase in performance brought by using functions from the standard library

Reviewer: We have not talked about ufcs or functions up until now.

Both **structs** and **classes** will be covered more in depth in a further lab.

### Quiz

What is the output of the following code?

```d
struct S
{
    int x;
}

class C
{
    int y;
}

void changeStruct(S s)
{
    s.x = 2;
}

void changeStruct(C c)
{
    c.y = 2;
}

void main()
{
    S s;
    C c = new C;

    changeStruct(s);
    changeStruct(c);

    import std.stdio : writeln;
    writeln(s.x);
    writeln(c.y);
}
```
    
- 0, 0
- 2, 0
- 2, 2
- 0, 2

### Exercise

Navigate to the struct-class directory. You will find 2 files implementing the same program. One uses a struct as a data container, while the other uses a class.

1. Compile both files and measure the run time of each program. How do you explain the differences?
2. In both situations, print the value of the field a0 after the loop ends. How do you explain the differences? How we can change the code for structs so both versions produce the same output?

> ## Functions

Reviewer: Maybe you could add the 3rd exercise from the first pool of exercises after this paragraph?

1. Imports may selectively specify which symbols are to be imported. For example, in the above code snippet, the full standard IO module of the standard library is imported, even though only the **printf** function is used.
This results in a degradation in compile time since there is a larger symbol pool that needs to be examined when trying to resolve a symbol.
In order to fix this, we can replace the blunt import with:
```d
import std.stdio : printf;
```

2. Imports may be used at any scope level.
```d
int main()
{
    int position = 7, c, n = 10;
    int[10] array = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

    for (c = position - 1; c < n - 1; c++)
        array[c] = array[c+1];

    // if the lines calling printf are deleted,
    // it is easier to spot the now useless import
    import std.stdio : printf;
    printf("Resultant array:\n");
    for (c = 0; c < n - 1; c++)
        printf("%d\n", array[c]);

    return 0;
}
```

### Exercise

C to D
Go to this [link](https://github.com/TheAlgorithms/C/tree/master/searching). You will find a series of searching algorithms, each implemented in C in its own file.

> - Choose one algorithm and port it to D with minimum modifications.

Author: maybe we can get rid of this?

> - Choose one algorithm and port it to D with minimum modifications.

Reviewer 2: Not sure. I think it's a good first exercise to just port something from C to D an see how easy it is (supposedly).

> - Choose one algorithm and port it to D with minimum modifications.

Reviewer: Yes, the exercise is good. I like the fact that they can choose. I suggest we leave this as is.

     
        a = b ~ c;       // a will be [1, 2, 3, 4, 5]
        a = b;           // a refers to b
        a = b ~ c[0..0]; // a refers to a copy of b
        a ~= c;          // equivalent to a = a ~ c;
    }
    ```
       
    Concatenation always creates a copy of its operands, even if one of the operands is a 0-length array.
    The operator `~=` does not always create a copy.
    For more information click [here](https://dlang.org/spec/arrays.html#resize).
    
3. Vector operations:

    ```d
    void main()
    {
        int[] a = [1, 2, 3];
        int[] b;
        int[] c;
     
        b[] = a[] + 4;         // b = [5, 6, 7]
        c[] = (a[] + 1) * b[]  // c = [10, 18, 28]
    }
    ```
    
    Many array operations, also known as vector operations, can be expressed at a high level rather than as a loop.
    A vector operation is indicated by the slice operator appearing as the left-hand side of an assignment or an op-assignment expression.
    The right-hand side can be an expression consisting either of an array slice of the same length and type as the left-hand side or a scalar expression of the element type of the left-hand side, in any combination.

>     Using the concepts of slicing and concatenation, we can modify the original example (that does the removal of an element from the array) so that the **for** loop is no longer necessary:

Reviewer: You could maybe ask them to do this.

>     Using the concepts of slicing and concatenation, we can modify the original example (that does the removal of an element from the array) so that the **for** loop is no longer necessary:

Reviewer: On a second thought. It would be better to leave this example and then ask them to improve the code from the first exercise by using slices.

    ```
    
    Overloading is done based on the **type of parameters**, not on the **return type**.
    
3. Default parameters:

    ```d
    void fun(int a, int b=8) {}
     
    void main()
    {
        fun(7);    // calls fun(7, 8)
        fun(2, 3); // calls fun(2, 3) 
    }
    ```
        
    A function may have any number of default parameters. If a default parameter is given, all following parameters must also have default values.
    
4. Auto return type:

    ```d
    auto fun()
    {
        return 7;
    }
     
    void main()
    {
        int b = fun();
        auto c = fun();   // works for variables also
    }
    ```
        
    Auto functions have their return type inferred based on the type of the return statements.
    Auto can be also used to infer the type of a variable declaration.
    
### Exercise

Navigate to the **voldemort** directory. Inspect the source file **voldermort.d**.
The declaration of struct **Result** is declared as **private** (nobody has access to it, except the members of the current file).
Move the declaration of the struct **Result** inside the **fun** function.
Compile the code. Does it compile? Why? Fix the issue.
    

> ## Unittests

Reviewer 2: IMO, `unittests` are really important. Add an exercise here. Start from a demo where some `unittests` fail and have the students fix them, then add a few new ones.

There can be any number of unit test functions in a module, including within struct and class declarations.
They are executed in lexical order.
Unit tests, when enabled, are run after all static initialization is complete and before the **main()** function is called.

```d
struct Sum
{
    int add(int x, int y) { return x + y; }
 
    unittest
    {
        Sum sum;
        assert(sum.add(3,4) == 7);
        assert(sum.add(-2,0) == -2);
    }
}
```
   
## Contract Programming

Contracts are a breakthrough technique to reduce the programming effort for large projects. Contracts are the concept of preconditions, postconditions and errors.

### Preconditions/Postconditions

The pre contracts specify the preconditions before a statement is executed. The most typical use of this would be in validating the parameters to a function.
The post contracts validate the result of the statement.
The most typical use of this would be in validating the return value of a function and of any side effects it has.
In D, pre contracts begin with in, and post contracts begin with out.
They come at the end of the function signature and before the opening brace of the function body.

> ```d

Reviewer 2: This snippet can be moved to a demo and extended to an exercise.

    return 0;
}
 
int fun(ref int a, int b)
in
{
    assert(a > 0);
    assert(b >= 0, "b cannot be negative!");
}
out (r)
{
    assert(r > 0, "return must be positive");
    assert(a != 0);
}
do
{
    // function body
    return 0;
}
```

The two functions are almost identical semantically.
The expressions in the first are lowered to contract blocks that look almost exactly like the second, except that a separate block is created for each expression in the first, thus avoiding shadowing variable names.

### Invariants

Invariants are used to specify characteristics of a class or struct that must always be true (except while executing a member function).
For example, a class representing a date might have an invariant that the day must be 1..31 and the hour must be 0..23:

```d
class Date
{
    int day;
    int hour;
 
    this(int d, int h)
    {
        day = d;
        hour = h;
    }
 
    invariant
    {
        assert(1 <= day && day <= 31);
        assert(0 <= hour && hour < 24, "hour out of bounds");
    }
}
```

For public or exported functions, the order of execution is:
1. preconditions
2. invariant
3. function body
4. invariant
5. postconditions

### Quiz

> What is the output of the following code?

Reviewer 2: In addition, "what is the output of this program" is not a very relevant quiz question in general. We don't want to teach students to be human compilers.

```d
import std.uni : toLower;

int fun(string s)
in (s == s.toLower())
{
    return -42;
}

string fun(int x)
out (; x >= 0)
{
    x *= -42;
    return "Hello World!";
}

void main()
{
    import std.stdio : writeln;

    writeln(fun("hello world!"));
    writeln(fun(-42));
}
```
- Runtime error at 1st call of fun()
- Runtime error at 2nd call of fun()
- hello world!
-42
- -42
Hello World!

### Exercise

> Navigate to the binary-search directory.

Reviewer 2: ```suggestion
Navigate to the `binary-search` directory.
```

> Inspect the source file binarySearch.d.

Reviewer 2: ```suggestion
Inspect the source file `binary_search.d.`
```

As the name implies, a binarySearch algorithm is implemented on integers.
Compile and run the file.

> - Write a unittest function that tests some corner cases. Are there any bugs in the algorithm implementation? If yes, fix them.

Reviewer 2: ```suggestion
- Write a unittest function that tests some corner cases.
Are there any bugs in the algorithm implementation?
If so, fix them.
```

> - The binarySearch function may be called with invalid data (for example: l = -1, r = -2). Write an in contract that halts the code execution in the case of invalid input (invalid values for l and r, the array is not sorted etc.)

Reviewer 2: ```suggestion
- The `binarySearch` function may be called with invalid data (for example: `l` = -1, `r` = -2).
Write an `in` contract that halts the code execution in the case of invalid input (invalid values for l and `r`, the array is not sorted etc.)
```

- Rewrite the binarySearch algorithm to make use of slices.

### Exercise

> Navigate to array-median element. Follow the directions in the skeleton in order to implement our very own string : int associative array.

Reviewer 2: ```suggestion
Navigate to the `array-median` folder.
Follow the directions in the skeleton in order to implement our very own `string : int` associative array.
```
